### PR TITLE
API Skeleton

### PR DESCRIPTION
**Pull request by jinyingtan**
In this PR:
- Added AuthAPI skeleton 
- Added CategoriesAPI skeleton
- Added WishesAPI skeleton
- Added DonationsAPI skeleton
- Added UsersAPI skeleton
- Added ReportsAPI skeleton
- Added ReviewsAPI skeleton
- Added NPOOrganizationsAPI skeleton
- Update Prettier to check all `.js` files
- Added `.prettierignore`

### CODE REVIEW

File: utils/api/wishes.js [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+   * @return {object} The wish info
+   */
+  async getWish(id) {}
+
+  /**
+   * Get wishes belonging to a NPO. Only return 21 results
+   * @param {string} npoID 
+   * @return {array} A list of wishes belonging to a NPO
+   */
+  async getNPOWishes(npoID) {}
```
Comment (jamessspanggg) [inline utils/api/wishes.js:55]: Is there a need to separate into `getNPOWishes` and `getNextBatchOfNPOWishes`? would just `getNextBatchOfNPOWishes` be suffice?

File: utils/api/wishes.js [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+  async create(title, description, categories) {}
+
+  /**
+   * Get the top X wishes belonging to a category that are pending, sorted by timestamp
+   * @param {string} category The category name 
+   * @param {number} limit The number of wishes to return
+   * @throws {FirestoreError}
+   * @returns {array} A list of X wishes sorted by timestamp
+   */
+  async getTopXWishes(category, limit) {}
```
Comment (jamessspanggg) [inline utils/api/wishes.js:23]: how about when there are no category involved? i.e. `all`

File: utils/api/wishes.js [hunk 3]
```diff
@@ -0,0 +1,95 @@
+class WishesAPI {
+  #categoryPendingWishesQuery;
```
[the remark this answers is not in the document]
Comment (jinyingtan) [reply utils/api/wishes.js:2]: pending wishes for a particular category. What would be a better name for it?

File: utils/api/wishes.js [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+   * Get the top X wishes belonging to a category that are pending, sorted by timestamp
+   * @param {string} category The category name 
+   * @param {number} limit The number of wishes to return
+   * @throws {FirestoreError}
+   * @returns {array} A list of X wishes sorted by timestamp
+   */
+  async getTopXWishes(category, limit) {}
+
+  /**
+   * Get pending wishes belonging to a category. Only return 21 results
```
[the remark this answers is not in the document]
Comment (jinyingtan) [reply utils/api/wishes.js:26]: So in the comments, just state that it will return `BATCH_SIZE`?

File: utils/api/wishes.js (hunk 1 above)
Comment (jinyingtan) [reply utils/api/wishes.js:55]: The reason why I have 2 different functions is that `getNPOWishes` will be used when the user enters the NPO's profile page and the `getNextBatchOfNPOWishes` would be used for the dynamic loading when the user scrolls the page. I'm trying to keep it consistent with the category wishes.

File: utils/api/wishes.js (hunk 2 above)
Comment (jinyingtan) [reply utils/api/wishes.js:23]: There's currently no use case for top wishes without category. This is mainly used on the main page (for the wishes in the top categories). Do we still want to add an API call for that?

File: utils/api/wishes.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (jamessspanggg) [reply utils/api/wishes.js:2]: pendingWishesForCategoryQuery?

File: utils/api/wishes.js (hunk 2 above)
Comment (jamessspanggg) [reply utils/api/wishes.js:23]: i see, then can leave it for now

File: utils/api/wishes.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (jamessspanggg) [reply utils/api/wishes.js:26]: yeah, or just name the methods getBatchOfPendingWishes should be intuitive that the returned result is of BATCH_SIZE

File: utils/api/wishes.js (hunk 1 above)
Comment (jamessspanggg) [reply utils/api/wishes.js:55]: so calling `getNPOWishes` will always get the first BATCH_SIZE of wishes? if this is the case, would `getInitialBatchOfNPOWishes` be better?

File: utils/api/wishes.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (jinyingtan) [reply utils/api/wishes.js:2]: Alright will make the rename

File: utils/api/wishes.js (hunk 1 above)
Comment (jinyingtan) [reply utils/api/wishes.js:55]: yes. Would it still be confusing to separate the query?

Comment (jamessspanggg) [reply utils/api/wishes.js:55]: i think its fine if it is renamed to `getInitialBatchOfNPOWishes`

File: utils/api/donations.js [hunk 5]
```diff
@@ -0,0 +1,153 @@
+class DonationsAPI {
+  #pendingDonationsForCategoryQuery;
+  #prevPendingDonationsForCategoryDoc;
+  #donorDonationsQuery;
+  #prevDonorDonationsDoc;
+  #donorDonationsByStatusQuery;
+  #prevDonorDonationsByStatusDoc;
+  #completedNPODonationsQuery;
+  #prevCompletedNPODonationsDoc;
```
Comment (kohchihao) [inline utils/api/donations.js:9]: From line 2-9 is not required as we do not want the `API` to store state. Please remove it.

File: utils/api/donations.js [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+  async search(text) {}
+
+  /**
+   * Get the top X pending donations belonging to a category, sorted by timestamp
+   * @param {string} category The category name
+   * @param {number} limit The number of donations to return
+   * @throws {FirestoreError}
+   * @returns {array} A list of X pending donations, sorted by timestamp
+   */
+  async getTopXPendingDonations(category, limit) {}
```
Comment (kohchihao) [inline utils/api/donations.js:53]: consider changing `X` to `N` and the param should be `(category, n)`

File: utils/api/wishes.js [hunk 7]
```diff
@@ -0,0 +1,142 @@
+class WishesAPI {
+  #pendingWishesForCategoryQuery;
+  #prevPendingWishesForCategoryDoc;
+  #npoWishesQuery;
+  #prevNPOWishesDoc;
+  #npoWishesByStatusQuery;
+  #prevNPOWishesByStatusDoc;
+  #completedDonorWishesQuery;
+  #prevCompletedDonorWishesDoc;
```
Comment (kohchihao) [inline utils/api/wishes.js:9]: similar to the `DonationsAPI` pls remove this

**Pull request merged**